cumulusci/core/flowrunner.py

```diff
@@ -0,0 +1,277 @@
+- Task default (i.e. `.tasks.TASKNAME.options`)
+- Flow definition task options (i.e. `.flows.FLOWNAME.steps.STEPNUM.options`)
+- Flow definition subflow options (i.e. `.flows.FLOWNAME.steps.STEPNUM.options.TASKNAME`)
+    see `dev_org_namespaced` for an example
+- Flow runtime (i.e. on the commandline)
+
+
+"""
+
+# we don't actually use this set of imports, they're just in type
+# comments, which require explicit runtime import when checking...
+try:
+    from typing import List
+except ImportError:
+    pass
+
+import copy
+import logging
+from collections import namedtuple
+from distutils.version import LooseVersion
+
+from cumulusci.core.exceptions import FlowConfigError, FlowInfiniteLoopError
+from cumulusci.core.utils import import_class
+
+
+class StepSpec(object):
+    """ simple namespace to describe what the flowrunner should do each step """
+
+    def __init__(
+        self, step_num, task_name, task_options, allow_failure=False, from_flow=None
+    ):
+        self.step_num = step_num
+        self.task_name = task_name
+        self.task_options = task_options
+        self.allow_failure = allow_failure
+        self.from_flow = from_flow
+
+    def __repr__(self):
+        return "<StepSpec {num}:{name} {cfg}>".format(
+            num=self.step_num, name=self.task_name, cfg=self.task_options
+        )
+
+
+StepResult = namedtuple("StepResult", ["step_num", "task_name", "task"])
+
+
+class NoOpStep(object):
+    """ Sentinel object used to indicate a no-op step. """
+
+    def __repr__(self):
+        return "NoOpStep"
+
+    def __str__(self):
+        return self.__repr__()
+
+
+class FlowRunner(object):
+    def __init__(
+        self, project_config, flow_config, org_config, options=None, skip=None
+    ):
+        self.project_config = project_config
+        self.flow_config = flow_config
+        self.org_config = org_config
+
+        # TODO: Support Runtime Options
+        if not options:
+            options = {}
+        self.runtime_options = options
+
+        if not skip:
+            skip = []
+        self.skip = skip
+
+        self.results = []
+
+        self.logger = self._init_logger()
+        self.steps = self._init_steps()  # type: List[StepSpec]
+
+    def _run_step(self, step):
+        # get task implementation class for task_name in step
+        task_config = copy.deepcopy(self.project_config.get_task(step.task_name))
+        task_class = import_class(task_config.class_path)
+
+        # TODO: Actually override options based on step_config
+        # TODO: Resolve ^^task_name.return_value style option syntax
+
+        task = task_class(
+            self.project_config,
+            task_config,
+            org_config=self.org_config,
+            name=step.task_name,
+            stepnum=step.step_num,
+            flow=self,
+        )
+        task()
+        return StepResult(step.step_num, step.task_name, task)
+
+    def run(self):
+        for step in self.steps:
```

Comment:
I think most of the logging and callback hooks can/should get relatively centralized to this method.